core/ipc/rpc/kafka/src/main/java/org/opennms/core/ipc/rpc/kafka/KafkaRpcClientFactory.java

```diff
@@ -0,0 +1,311 @@
+                }
+                LOG.debug("RPC Request {} sent to minion at location {}", request.toString(), request.getLocation());
+                return future;
+            }
+        };
+
+    }
+
+    public void setLocation(String location) {
+        this.location = location;
+    }
+
+    public void start() {
+        // Set the defaults
+        kafkaConfig.clear();
+        kafkaConfig.put("group.id", SystemInfoUtils.getInstanceId());
+        kafkaConfig.put("enable.auto.commit", "true");
+        kafkaConfig.put("key.deserializer", StringDeserializer.class.getCanonicalName());
+        kafkaConfig.put("value.deserializer", ByteArrayDeserializer.class.getCanonicalName());
+        kafkaConfig.put("auto.commit.interval.ms", "1000");
+        kafkaConfig.put("key.serializer", StringSerializer.class.getCanonicalName());
+        kafkaConfig.put("value.serializer", ByteArraySerializer.class.getCanonicalName());
+        // This is needed to refresh topics metadata so that topics are added dynamically to consumer.
+        kafkaConfig.put("metadata.max.age.ms", TOPICS_METADATA_MAX_AGE);
+        kafkaConfig.put("auto.offset.reset", "earliest");
+
+        // Find all of the system properties that start with 'org.opennms.core.ipc.rpc.kafka.' and add them to the config. 
+        // See https://kafka.apache.org/10/documentation.html#newconsumerconfigs for the list of supported properties
+        for (Entry<Object, Object> entry : System.getProperties().entrySet()) {
+            final Object keyAsObject = entry.getKey();
+            if (keyAsObject == null || !(keyAsObject instanceof String)) {
+                continue;
+            }
+            final String key = (String) keyAsObject;
+
+            if (key.length() > KAFKA_CONFIG_SYS_PROP_PREFIX.length() && key.startsWith(KAFKA_CONFIG_SYS_PROP_PREFIX)) {
+                final String kafkaConfigKey = key.substring(KAFKA_CONFIG_SYS_PROP_PREFIX.length());
+                kafkaConfig.put(kafkaConfigKey, entry.getValue());
+            }
+        }
+        producer = new KafkaProducer<>(kafkaConfig);
+        // Start consumer which handles all the responses.
+        KafkaConsumer<String, byte[]> kafkaConsumer = new KafkaConsumer<>(kafkaConfig);
+        kafkaConsumerRunner = new KafkaConsumerRunner(kafkaConsumer);
+        executor.execute(kafkaConsumerRunner);
+        //Start a new thread which handles timeouts from delayQueue and calls response callback.
+        final ThreadFactory threadFactory = new ThreadFactoryBuilder().setNameFormat("rpc-client-timeout-tracker-%d")
+                .build();
+        Executors.newSingleThreadExecutor(threadFactory).execute(() -> {
+            while (true) {
+                try {
+                    ResponseCallback responseCb = delayQueue.take();
+                    responseCb.sendResponse(null);
+                } catch (InterruptedException e) {
+                    // exit, nothing to do
+                }
```

Comment:
I'd also add a generic `catch(Exception e)`  clause here - if an exception is throws while issuing a single callback we don't want to thread to abort and stop processing the queue.

Reply:
sure, makes sense. I will add that